rfcs/text/0000-query-by-metadata-schema-versions.md

```diff
@@ -0,0 +1,202 @@
+
+# RFC Name
+
+Querying DSS by Metadata Schema Version(s)
+
+## Summary
+
+This RFC details a design to allow downstream components to modify their subscription to the DSS to be only notified
+about bundles containing metadata within a specified schema version range. The selection of a solution is required to
+unblock the metadata schema integration testing detailed here.
+
+## Author(s)
+
+[Arathi Mani](mailto:arathi.mani@chanzuckerberg.com)
+
+## Shepherd
+
+[Matt Weiden](mailto:mweiden@chanzuckerberg.com)
+
+## Motivation
+
+Today in the DCP when metadata schema makes change, we require that the entire system, including all independent
+downstream components (i.e. Expression Matrix service, Secondary Analysis, Query service, Data Browser, etc.) are
+compatible with the changes before we allow the metadata team to push their changes through. This causes extreme
+slowness in metadata schema updates which is non-ideal due to the fact that schema changes are frequent and prevents new
+data from being added to the DCP as the new data is likely to adhere to the changed metadata schema.
+
+In order to enable more frequent metadata schema change-related pushes to production, a [metadata schema integration 
+test design](https://docs.google.com/document/d/18C_Wbrr4Huc8yxQXSdNbzeGk0glBjkfq0MGnozwlIB8/edit) has been proposed. 
+One part of design is to require downstream components to modify their production subscriptions to the Data Store 
+Service (DSS) such that they only receive notifications about bundles with which they are compatible. This will allow
+wranglers to submit new data to the DSS without having to wait for compatibility of the whole system and also allows
+downstream components to receive data independent of whether another component is ready as well. Due to the fact that
+there is currently a metadata schema freeze disallowing any changes, major, minor, or patch, to be pushed to production,
+this document simply designs the minimum system required to unblock wranglers from submitting data and protecting
+downstream components (and therefore the entire DCP) from breaking due to incompatibility.
+
+Note: the subscription modifications will only exist in the production environment. Once the metadata schema integration
+tests are setup, subscriptions will be modified to remove any restrictions on metadata schema versions in order to
+eagerly surface incompatibilities.
+
+
+### User Stories
+
+*Share the [User Stories](https://www.mountaingoatsoftware.com/agile/user-stories) motivating this RFC.*
+* As a data wrangle, I would like to be able to push new data adhering to the latest metadata schema without having to
+wait for every single component to be compatible with the new schema versions.
+* As a DCP developer working on a downstream component (e.g. Matrix Service, Secondary Analysis, etc.) I would like to
+receive notification about new data adhering to new metadata schema versions as soon as I am ready without having to
```

Comment:
Maybe more of a question for the metadata team, but as this RFC outlines a solution for downstream services to control updates to specific metadata versions, what is the best venue for developers to stay up-to-date with the latest metadata schema versions and, to their discretion, respond and update their services to support newer versions?

Reply:
Great question! This is definitely a discussion worth having.. 

From my perspective, once the metadata schema integration tests are implemented, there will be notifications sent out on "how your component broke with the latest schema changes." 

But a general digest might also make sense. @simonjupp , @malloryfreeberg , @morrisonnorman ... any thoughts?

Reply:
We have tried various tactics for communicating schema changes: Posting "notes" on Slack (which seemed people rarely read and were a 100% manual thing to maintain); Integration of metadata schema PRs with the #hca-metadata-int Slack channel (what we do now, don't know if people find it useful); Tagging individual people in tickets/PRs for changes in fields that we know they hard-code against (not sustainable; requires too much knowledge). 

Would love for the DCP to find a way to improve transparency of schema changes without adding manual work for schema developers. Do any other components communicate their code changes to the rest of the DCP? If so, does their communication system work? Maybe we can adapt things that work well for other teams :)

Reply:
Great response Mallory! I think today the DCP-wide tests that occur in integration, often surface errors between components and the metadata schema integration tests will help with that. We could try that out first and see if we need anything additional once that work is completed!

Reply:
Resolving conversation for now.. please feel free to reopen if everything is not addressed!